designs/disruption-controls-by-action.md

```diff
@@ -0,0 +1,188 @@
+# Disruption Controls By Action 
+## User Scenarios 
+1. I as a user would like to ensure upgrades only occur during business hours or stricter windows. I would also like to not restrict the cost savings of consolidation when blocking upgrade via drift.
+2. As a HFT firm, I have specific hours where I need all my compute and I can't tolerate scale down requests for consolidation, but outside of operational hours its fine to scale down. 
+## Desired Behavior + Requirements 
+1. User needs to be able to define an action and a budget/budgets for that action. 
+2. Actions that should be supported should by any disruption actions effected by the current Budgets implementation. (Consolidation, Emptiness, Expiration, Drift) 
+3. Budgets should still support the option to set a given behavior for all disruption actions. If action is unspecified we want to just assume its mean't to apply to all actions. 
+
+## API Design
+### Approach A: Add an action field to disruption Budgets 
+This approach outlines a simple api change to the betav1 nodepool api to allow disruption budgets to specify a disruption action. 
+### Proposed Spec
+Add a simple field "action" is proposed to be added to the budgets. 
+```go
+// Budget defines when Karpenter will restrict the
+// number of Node Claims that can be terminating simultaneously.
+type Budget struct {
+	// Action defines the disruption action that this particular disruption budget applies to. 
+	Action DisruptionAction `json:"action,omitempty" hash:"ignore"`
```

Comment:
So, I would at least want this to be an array, so the budget could be applied to multiple actions together. Feels not as great if you have to redefine the same budget for 3+ actions.

Another possible option gaining some of the value from Approach B is by adding action profiles within the Budget struct instead as in:
`ConsolidationProfile`
`DriftProfile`
`ExpirationProfile`
`EmptinessProfile`

However, that might be adding far too much bulk to this API, and defining something like this elsewhere might be desired if we want it.

Reply:
> So, I would at least want this to be an array, so the budget could be applied to multiple actions together. Feels not as great if you have to redefine the same budget for 3+ actions.
Great point

Reply:
Allthough one should be able to specify a budget through the "" action or for all remaining actions. It does bring additional complexity to resolving the budgets ,and we only have 4 actions today so seems a bit excessive. 

thoughts?

Reply:
After reading over `Q: How should Karpenter handle the default or undefined action case?` realizing that an array form has the same question of is it a shared budget? 

I was assuming the array idea would be the same as defining multiple budgets (one for each of the ones in the array), but that gives potentially different behavior than the "All"/default option.

Reply:
Thinking more about it, extending the funcitonality to drift and other things means we can have many more than 4 methods, and this evolves nicely.